File: four-sum/FourSum.swift
import Foundation

class FourSum {
    public func fourSum(array: inout [Int], targetSum: Int) -> [[Int]] {
        array.sort{ $0 < $1}
        var results: [[Int]] = []
        for i in stride(from: 0, to: array.count, by: 1) {
            for j in stride(from: i + 1, to: array.count, by: 1) {
                var left = j + 1
                var right = array.count - 1
                while(left < right) {
                    var sum = array[i] + array[j] + array[left] + array[right]
                    if(sum == targetSum) {
                        results.append([ array[i], array[j], array[left], array[right]])
                        left += 1; right -= 1;
                    } else if (sum < targetSum) {
                        left += 1
                    } else if (sum > targetSum) {
                        right -= 1
                    }
                }
            }
        }
        return results
    }
}

File: four-sum/FourSum.java
import java.util.*;

public class FourSum {
    public static List<Integer[]> fourSum(int[] array, int targetSum) {
        Arrays.sort(array);
        List<Integer[]> results = new ArrayList<>();
        for(int i = 0; i < array.length; i++) {
            for(int j = i + 1; j < array.length; j++) {
                int left = j + 1;
                int right = array.length - 1;
                while(left < right) {
                    int sum = array[i] + array[j] + array[left] + array[right];
                    if(sum == targetSum) {
                        results.add(new Integer[] { array[i], array[j], array[left], array[right] });
                        left++; right--;
                    } else if (sum < targetSum) {
                        left++;
                    } else {
                        right--;
                    }
                }
            }
        }
        return results;
    }
}